Optimize string building and interning

ladipro opened the pull request:
### Overview

This PR addresses the following issues:

1. The current version of string interning takes a custom `IInternable` interface as its input. Since most implementations of the interface are structs, the code is duplicated at run-time using generics to prevent boxing: `void SomeFunction<T>(T internable) where T : IInternable`.
2. It's been pointed out that `IInternable` is functionally equivalent to `ReadOnlySpan<char>`. `IInternable` needs a separate and manually authored implementation for each kind of input, whereas `ReadOnlySpan<char>` comes from BCL for free. We are currently missing a wrapper over `char *, length`, for example.
3. One of the string-like datatypes implementing `IInternable` is our `StringBuilder` wrapper. Interning iterates over the candidate string linearly, hashing its contents and comparing it with already interned strings. For `StringBuilder` this operation is O(N^2) because of its internal representation and the lack of an efficient enumerator.
4. Most uses of `StringBuilder` in MSBuild are "append-only" where the resulting string is composed by concatenating a series of string fragments representable with `ReadOnlySpan<char>` or `ReadOnlyMemory<char>`. This usage pattern allocates 2x the length of the resulting string - one copy lives in the `StringBuilder`'s internal char array(s) and the other is created as part of the `ToString()` call.

Introduced in this change is a new type named `SpanBasedStringBuilder`, serving as an efficient `StringBuilder`-like builder. Also provided is a static class named `Strings` with functionality to intern an arbitrary `ReadOnlySpan<char>`.

a) Interning read-only character spans
Makes it possible to intern a string, substring, character array/sub-array, char *, or any other `ReadOnlySpan<char>`. The advantage over first creating the desired `System.String` and then interning it is that it does away with the ephemeral `System.String` allocation in the case where the string already is in the intern table.

```C#
var str = Strings.WeakIntern("semicolon;delimited".AsSpan(0, 8));
Console.WriteLine(str); // prints "semicolon"
```

Memory characteristics of the above snippet:
- The WeakIntern() call does not allocate if the string "semicolon" already exists in the intern table,
- The call allocates the string and a small overhead if the string is not yet in the intern table.

b) An internable `StringBuilder`.
Compensates for the fact that `System.Text.StringBuilder` cannot be represented with one `ReadOnlySpan<char>` and that it does not provide a linear-time character enumeration. The usage pattern is similar to that of a `StringBuilder` but it does not allocate O(N) bytes where N is the intermediate string length, but rather allocates only spans describing its constituent pieces. Note that in degenerated cases this may be worse than O(N) so make sure it's used only where it's actually helping.

```C#
using var str = Strings.GetSpanBasedStringBuilder();
str.Append("semicolon");
str.Append(";");
str.Append("delimited");
Console.WriteLine(str.ToString()); // prints "semicolon;delimited"
```

Memory characteristics of the above snippet:
- `SpanBasedStringBuilder` instances are pooled so in the expected case the GetSpanBasedStringBuilder() call does not allocate,
- `SpanBasedStringBuilder` holds a List of span descriptors on the GC heap, the size is O(S) where S is the number of spans (in this example 3),
- The `Append()` calls do not allocate memory unless they need to resize the List of span descriptors,
- The `ToString()` call does not allocate the string "semicolon;delimited" if it already exists in the intern table.

### Code structure
- `ReuseableStringBuilder` is simplified as it no longer supports interning (going forward most of its uses should be replaced with `SpanBasedStringBuilder`),
- All uses of `ReusableStringBuilder` that used interning are converted to `SpanBasedStringBuilder`,
- All string/substring/char array interning is converted to `Strings.WeakIntern()`,
- `IInternable`, `WeakStringCache`, `OpportunisticIntern` are all deleted and the functionality moved to a new library called `StringTools`,
- `StringTools` is built for three targets: .NET Standard, .NET Framework 4, and .NET Framework 3.5 to be used in MSBuildTaskHost, and as usual, the 3.5 implementation is rather simplified as we're fine with sub-optimal performance,

Note that no effort has been made to convert other uses of `ReuseableStringBuilder` and `StringBuilder` to `SpanBasedStringBuilder`. The expectation is that there are many opportunities to eliminate ephemeral allocations and it should be done in separate PRs.

### To-do
- [x] Run an experimental VS insertion
- [x] Measure performance impact by profiling project evaluation

ladipro commented:
Checks failed due to #5506

ladipro commented:
/azp run

cdmihai commented:
> <Project Sdk="Microsoft.NET.Sdk">

Adding a new MSBuild assembly will cause a bit of propagation pain to all the tools that have hard coded lists with all the msbuild assemblies, like Microsoft.Build.Locator.

---
Refers to: src/StringTools/StringTools.csproj:1 in e1846cc. [](commit_id = e1846cc9030617acf5ed1af077bd4820b794f30f, deletion_comment = False)

cdmihai commented on src/Build/BackEnd/BuildManager/BuildManager.cs:
```diff
@@ -405,7 +405,7 @@ public void BeginBuild(BuildParameters parameters)
                 if (BuildParameters.DumpOpportunisticInternStats)
                 {
-                    OpportunisticIntern.Instance.EnableStatisticsGathering();
+                    StringTools.StringTools.EnableDiagnostics();
```
nit: import StringTools namespace.

ladipro replied:
Done. Also renamed the `StringTools` class because a type named `StringTools` in a namespace named `StringTools` is only causing issues.

cdmihai commented on src/Build/Construction/Solution/SolutionProjectGenerator.cs:
```diff
@@ -1097,8 +1100,11 @@ private static string GetConfigurationAndPlatformPropertiesString(ProjectConfigu
         {
-            string directProjectProperties = OpportunisticIntern.InternStringIfPossible(String.Join(";", GetConfigurationAndPlatformPropertiesString(projectConfiguration), SolutionProperties));
-            return directProjectProperties;
+            using RopeBuilder directProjectProperties = ST.GetRopeBuilder();
```
Didn't reach StringTools yet, but `RopeBuilder` does not sound descriptive at all. Is it some established pattern / algorithm that I never head about? By its name I assume it takes multiple sequencey things and somehow weaves them together :)

cdmihai replied:
On a darker note, it's also indicative of how we feel sometimes when reading the msbuild code base :)

---
In reply to: [486037469](https://github.com/dotnet/msbuild/pull/5663#discussion_r486037469) [](ancestors = 486037469)

drewnoakes replied:
Haven't looked but guessing it's this: https://en.wikipedia.org/wiki/Rope_(data_structure)

ladipro replied:
Yes, it's a rope, although a simple one since we don't need operations that would make it a tree. We basically just append strings to the end. I want to have the API (informally) reviewed before merging so it is possible that we'll end up with a different name.

ladipro replied:
Renamed to `SpanBasedStringBuilder`.

cdmihai commented on src/Build/Definition/Project.cs:
```diff
@@ -109,6 +109,7 @@ internal Project(ProjectCollection projectCollection, ProjectLink link)
             ProjectCollection = projectCollection;
             implementationInternal = new ProjectLinkInternalNotImplemented();
             implementation = link;
+            StringToolsHost.Initialize();
```
This could be called in a static Project constructor right? To make it more clear that it's a static, once in a process thing.

ladipro replied:
Indeed, I've moved it to the static ctor, thanks!

cdmihai commented on src/Package/MSBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec:
```diff
@@ -32,6 +32,7 @@
     <file src="$X86BinPath$/System.Memory.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Text.Json.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/Microsoft.Bcl.AsyncInterfaces.dll" target="v15.0/bin" />
+    <file src="$X86BinPath$/Microsoft.Extensions.ObjectPool.dll" target="v15.0/bin" />
```
Don't you also have to add the StringTools assembly in here?

ladipro replied:
Thanks, fixed! I have yet to verify the whole thing with an exp insertion. It is likely that more packaging-related issues will surface.

cdmihai commented on src/Package/MSBuild.VSSetup/files.swr:
```diff
@@ -40,6 +40,7 @@ folder InstallDir:\MSBuild\Current\Bin
   file source=$(X86BinPath)System.Memory.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Text.Json.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)Microsoft.Bcl.AsyncInterfaces.dll vs.file.ngenArchitecture=all
+  file source=$(X86BinPath)Microsoft.Extensions.ObjectPool.dll vs.file.ngenArchitecture=all
```
Add StringTools as well?

ladipro replied:
Thanks, done!

cdmihai commented on src/Shared/Constants.cs:
```diff
@@ -60,8 +60,6 @@ internal static class MSBuildConstants
         internal const string CurrentToolsVersion = "Current";
 
-        // if you change the key also change the following clones
-        // Microsoft.Build.OpportunisticIntern.BucketedPrioritizedStringList.TryIntern
```
I'd leave the comment and update it with the new location

cdmihai replied:
Ah, comment no longer relevant

---
In reply to: [486044656](https://github.com/dotnet/msbuild/pull/5663#discussion_r486044656) [](ancestors = 486044656)

cdmihai commented on src/Shared/InterningBinaryReader.cs:
```diff
@@ -40,15 +40,15 @@ internal class InterningBinaryReader : BinaryReader
         /// </summary>
         private InterningBinaryReader(Stream input, Buffer buffer)
-            : base(input, buffer.Encoding)
+            : base(input, Encoding.UTF8)
```
Why the hardcoded encoding?

ladipro replied:
It was already hardcoded before, just in a more convoluted way, presumably due to a perf concern. I've simplified the code a tiny bit.

ladipro commented:
> Adding a new MSBuild assembly will cause a bit of propagation pain to all the tools that have hard coded lists with all the msbuild assemblies, like Microsoft.Build.Locator.

We have added several dependent .dll's recently so I was under the assumption that it is not a super disruptive change.
- System.Text.Json.dll for solution filter support,
- StreamJsonRpc.dll & friends for modern IPC serialization.

In case of this library it is unfortunately not possible to statically link it into MSBuild as sharing it among all components in the process is an explicit goal. I am curious to learn more about the implications of adding a dependency.

cdmihai commented:
>I am curious to learn more about the implications of adding a dependency.

As far as I'm aware:
- code that hard codes msbuild assemblies will potentially need updating (like [here](https://github.com/microsoft/MSBuildLocator/blob/ed4fce73f2f946c08f7ecf6f2a9cb64b056003f5/src/MSBuildLocator/MSBuildLocator.cs#L22-L28) or [here](https://github.com/microsoft/BuildXL/blob/45407251608694485f9e22ff9e3fbb3d17449381/Public/Src/Tools/Tool.MsBuildGraphBuilder/MsBuildAssemblyLoader.cs#L61-L72))
- [app.config](https://github.com/dotnet/msbuild/blob/c7790e1ed37c2e6cb4fc38e76c06ed1fb15d3148/src/MSBuild/app.config#L15-L42) probably needs a new redirect
- update all the packages

ladipro commented:
@danmosemsft may I ask you to take a look at the StringTools library introduced in this PR? This is the library to be published and used by multiple components running in the devenv process that we recently agreed would initially live in the msbuild repo. Thank you!

danmosemsft commented:
@ladipro thanks for sharing. I will try to look but maybe not today. 

It sounds like the idea is that this would be a generic interning service for any managed code in VS? 

As you know there are countless strategies one could use, from the PR it's only possible to review for correctness and not be confident about the approach. I would expect you plan to measure (in MSBuild and VS) and iterate up the wazoo. Including with non MSBuild consumers.

cc @jkotas in case he is interested in your strategies. Glad you are tackling this!

danmosemsft commented:
This is the kind of thing @AArnott may be interested in.

danmosemsft commented on src/StringTools/RopeBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+    /// The <see cref="ToString"/> method tries to intern the resulting string without even allocating it if it's already interned.
+    /// Use <see cref="StringTools.GetRopeBuilder"/> to take advantage of pooling to eliminate allocation overhead of this class.
+    /// </remarks>
+    public class RopeBuilder : IDisposable
```
As I understand it, Ropes is a well understood structure that is a balanced tree. I would avoid using the term for a list of spans. SpanCollection? SpanOfCharList?

ladipro replied:
Renamed to `SpanBasedStringBuilder`. The name is a bit longer but I want to make it clear that it is supposed to be functionally analogous to `StringBuilder`.

danmosemsft commented on src/StringTools/RopeBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+            {
+                ReadOnlySpan<char> span = _spans[spanIdx].Span;
+                int i = span.Length - 1;
+                while (i >= 0 && char.IsWhiteSpace(span[i]))
```
You can use the built in Trim() method. eg https://apisof.net/catalog/System.MemoryExtensions.Trim%3CT%3E(Span%3CT%3E,T)

ladipro replied:
`Trim()` is not available on the Framework version of `ReadOnlyMemory` unfortunately and running `.Span.Trim()` and then slicing the `ReadOnlyMemory` according to the trimmed span is almost twice as slow compared to what I have here. Unless there's a better way that has comparable performance I'd say it's better to keep this implementation but I'll defer to your opinion.

danmosemsft commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,232 @@
+        /// via the <see cref="InternableString"/> representation and can override the default behavior by returning a custom
+        /// string.
+        /// </remarks>
+        public static void RegisterStringInterningCallback(TryInternStringDelegate callback)
```
This callback scheme is a complexity it would be good to avoid. The decision to hard code a lookup for short well known strings was based on measurements a long time ago, and also in a context where there was no callback involved. I suggest to rip it out and do your measurements with no strings propopulated. You can look at the stats dump you have built in to see which strings are very common (they may have changed since this list was chosen). Then you can decide whether the data suggests they need special handling. If they do that you might not choose the same callback-and-switch scheme but something else.

ladipro replied:
That's a great point. The callback used by MSBuild has both "positive" and "negative" strings and I was more worried about the latter, i.e. strings that must not be interned. It doesn't look like we actually depend on this in any way, though, and it's probably just a left-over from a logic removed long ago.

I'll take detailed measurements and will get back on the value of hardcoded strings. Thank you!

ladipro replied:
I have removed the callback, you were right that hardcoded strings were not helping. Thank you!

cdmihai commented on src/Shared/ReuseableStringBuilder.cs:
```diff
@@ -17,19 +17,13 @@ namespace Microsoft.Build.Shared
     /// <remarks>
     /// You can add any properties or methods on the real StringBuilder that are needed.
     /// </remarks>
-    internal sealed class ReuseableStringBuilder : IDisposable, IInternable
```
Ideally this class should be subsumed by the RopeBuilder right? Worth adding a class comment saying that?

ladipro replied:
Done, thanks!

ladipro commented:
> It sounds like the idea is that this would be a generic interning service for any managed code in VS?

Yes. Still debating whether it needs to be a formal VS "service" but it's going to be at least a regular library. In the MSBuild (and I'm guessing also Roslyn) case, the same code runs out-of-proc as well as in-proc so it's convenient to simply link against a library and then automatically get process-wide string sharing, be it the devenv process or an out-of-proc worker.

>As you know there are countless strategies one could use, from the PR it's only possible to review for correctness and not be confident about the approach. I would expect you plan to measure (in MSBuild and VS) and iterate up the wazoo. Including with non MSBuild consumers.

That's the idea. I'm starting with MSBuild and will then consume the library from other places as well. Nothing is set in stone and there are likely going to be adjustments, although the core `ReadOnlySpan<char> -> string` interning function is likely going to stay the way it is - it is super simple and should cover most scenarios.

cdmihai commented on src/StringTools/SpanBasedStringBuilder.Simple.cs:
```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        public SpanBasedStringBuilder(int capacity = 4)
+        {
+            _builder = new StringBuilder(capacity * 128);
```
nit: 128 is the expected span length I assume. Make constant? Same for 4

cdmihai replied:
Nevermind, this is the net35 implementation.

cdmihai commented on src/StringTools/SpanBasedStringBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+                {
+                    i++;
+                }
+                if (i > 0)
```
How can i be negative here?

cdmihai commented on src/StringTools/SpanBasedStringBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+                if (i > 0)
+                {
+                    _spans[spanIdx] = _spans[spanIdx].Slice(i);
+                    Length -= i;
```
Confused as to how this method works. I'd expect:
- if the first X spans are empty, it removes them from the beginning of the span list. If all spans contain whitespace it leaves the span list empty
- it slices the first non empty span

cdmihai replied:
Same goes for TrimEnd.

ladipro replied:
It depends on the way `SpanBasedStringBuilder` is used. For the typical "call `Append` a few times followed by `ToString`" the overhead of removing empty spans may not be worth the savings because there's just a few passes over the string that can benefit from this and then the entire thing is released. I'll experiment with this suggestion using a microbenchmark.

cdmihai commented on src/StringTools/InternableString.cs:
```diff
@@ -0,0 +1,343 @@
+    /// The structure is public because it's exposed via the <see cref="TryInternStringDelegate"/> callback. It's a mutable struct and should
+    /// generally not be used directly from outside of the library.
+    /// </remarks>
+    public ref struct InternableString
```
`IInternable` was able to represent any string like structure whereas `InternableString` only works with spans. The assumption is that any string like structure that msbuild wants to intern should be convertible to a span, which as far as I can tell is a pretty good assumption.

Forgind commented on src/StringTools/InternableString.cs:
```diff
@@ -0,0 +1,343 @@
+    /// generally not be used directly from outside of the library.
+    /// </remarks>
+    public ref struct InternableString
+    {
```
I'm a little confused about when someone should use InternableString vs. SpanBasedStringBuilder. Since they have some similarities in their implementation, some of my comments there might also apply here.

ladipro replied:
You're right, it was confusing. The only reason `InternableString` was exposed was to pass it to the intern callback. Now that the callback is gone, the type can be internal and not part of the public surface. Thanks!

Forgind commented on src/StringTools/OpportunisticIntern.cs:
```diff
@@ -0,0 +1,295 @@
+    /// <summary>
+    /// Interns strings by holding weak references.
+    /// </summary>
+    internal sealed class OpportunisticIntern : IDisposable
```
Should this class be static?

ladipro replied:
This code is gone now. I believe it was implemented like this (state in instance fields + a static `Instance` getter) for testability, i.e. all state can be easily reset by re-creating the singleton instance.

Forgind commented on src/StringTools/OpportunisticIntern.cs:
```diff
@@ -0,0 +1,295 @@
+                if (_gatherStatistics)
+                {
+                    result.AppendLine(string.Format("\n{0}{1}{0}", new string('=', 41 - (title.Length / 2)), title));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Hardcoded Hits", _hardcodedInternHits, "hits"));
```
Should all these strings be localized?

ladipro replied:
For debug output behind an undocumented environment variable I vote for no. There are many places in the codebase producing non-localized debug output, for example `DumpSchedulerState`.

Forgind commented on src/StringTools/OpportunisticIntern.cs:
```diff
@@ -0,0 +1,295 @@
+                    IEnumerable<string> topMissingHardcodedString =
+                        _missedHardcodedStrings
+                        .OrderByDescending(kv => kv.Value * kv.Key.Length)
+                        .Take(15)
```
What if there are under 15?

And where did 15 come from?

And should the Where come before the Take?

ladipro replied:
> What if there are under 15?

Then everything will be returned. From `Enumerable.Take<>` docs:
_If count exceeds the number of elements in source, all elements of source are returned._

> And where did 15 come from?

Just an arbitrary number to make the output reasonable in size. I copied this from the old code.

> And should the Where come before the Take?

Yes, it makes more sense that way. Fixed!

Forgind commented on src/StringTools/OpportunisticIntern.cs:
```diff
@@ -0,0 +1,295 @@
+                            break;
+                    }
+
+                    if (internResult != InternResult.MatchedHardcodedString && internResult != InternResult.RejectedFromInterning)
```
tiny nit:
```suggestion
                    if (internResult == InternResult.FoundInWeakStringCache || internResult == InternResult.AddedToWeakStringCache)
```

ladipro replied:
Done thanks!

Forgind commented on src/StringTools/OpportunisticIntern.cs:
```diff
@@ -0,0 +1,295 @@
+            /// </summary>
+            private string InternWithStatistics(ref InternableString candidate)
+            {
+                lock (_missedHardcodedStrings)
```
Does the lock need to be over this whole section or just the if statement below?

ladipro replied:
This code has moved but I believe that your question still applies. In the current version of the code the lock protects several fields: `_regularInternHits`, `_regularInternMisses`, `_internCallCountsByString`, `_internEliminatedStrings`, `_internEliminatedChars`. Integer fields could be accessed lock-free with interlocked operations but given that we're also mutating a dictionary, it's easier to put everything under one lock. The code is used only for gathering statistics for debugging so it's not super perf sensitive.

Forgind commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,233 @@
+            /// </summary>
+            public bool Return(SpanBasedStringBuilder stringBuilder)
+            {
+                if (stringBuilder.Capacity <= MAX_RETAINED_BUILDER_CAPACITY)
```
Why this check?

ladipro replied:
The idea was to not return the builder to the pool if it was excessively large. The idea is that if the object is unlikely to be needed again, it is better to throw it away than to keep it forever. I have removed this check, thanks!

Forgind commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,233 @@
+        /// <summary>
+        /// An array of callbacks to be called for each string being interned.
+        /// </summary>
+        private static volatile TryInternStringDelegate[] s_internStringCallbacks = new TryInternStringDelegate[0];
```
Would you mind explaining every part of this?

Also, nit:
```suggestion
        private static volatile TryInternStringDelegate[] s_internStringCallbacks = Array.Empty<TryInterStringDelegate>();
```

Forgind replied:
Actually, should this be an array? I didn't see anywhere that you used it without looking at every element, and you sometimes add to the end. Maybe a LinkedList? Arrays have less overhead, but a LinkedList might be faster...

ladipro replied:
I wanted a data structure that can be read in a lock-free fashion. Since writes are expected to be super rare, I went with the simplest type - an immutable array. Readers get a reference to the array and read it lock-free because no writer is modifying the array once published into the static. Only writers have to synchronize between them so they don't step on each other toes. Done with a lock here, could be a loop with an interlocked exchange as well. Also note the `volatile` modifier which ensures that the contents of the array is visible to other threads when a writer does `s_internStringCallbacks =`.

But maybe all this is moot if, as Dan suggested, we don't need this callback mechanism at all. I am going to measure this shortly.

Forgind commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,233 @@
+        /// A pool of SpanBasedStringBuilders as we don't want to be allocating every time a new one is requested.
+        /// </summary>
+        private static DefaultObjectPool<SpanBasedStringBuilder> s_stringBuilderPool =
+            new DefaultObjectPool<SpanBasedStringBuilder>(new PooledObjectPolicy(), Environment.ProcessorCount);
```
Environment.ProcessorCount is the often-unreliable processor count, right?

ladipro replied:
I have ultimately decided against using `ObjectPool` in the first version of the library. The builders are now pooled in a `[ThreadStatic]` field, so one per thread which is makes them fast to get and return, and perfectly fine for MSBuild (suspecting ought to be enough for everybody).

`Environment.ProcessorCount` was meant to be a heuristic on the number of concurrent users but it makes potentially incorrect assumptions.

Forgind commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,233 @@
+        /// </summary>
+        internal static void ResetForTests()
+        {
+            s_internStringCallbacks = new TryInternStringDelegate[0];
```
tiny nit:
```suggestion
            s_internStringCallbacks = Array.Empty<TryInternStringDelegate>();
```

ladipro replied:
Can't do this because the code has to compile against .NET Framework 3.5 where `Array.Empty<>` does not exist yet.

Forgind commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,233 @@
+            // We can read s_internStringCallbacks lock-less because it's declared volatile and RegisterStringInterningCallback
+            // and UnregisterStringInterningCallback write the field after initializing its content.
+            TryInternStringDelegate[] callbacks = s_internStringCallbacks;
+            foreach (TryInternStringDelegate callback in callbacks)
+            {
+                if (callback(ref candidate, out interned))
+                {
+                    return true;
+                }
+            }
```
```suggestion
            return callbacks.Any(callback => TryInternStringDelegate(callback, out interned));
```
? Might change the value of interned, so you'd have to account for that when using it. (You already do, for now.)

ladipro replied:
This code is gone, no more hardcoded strings.

Forgind reviewed (commented):
I'm only about halfway through my review. Still need to look at tests and most of the places you used it.

Sorry if there are duplicates of others' comments; I started this a while ago and haven't always been keeping up since.

Forgind commented on src/StringTools.UnitTests/OpportunisticIntern_Tests.cs:
```diff
@@ -0,0 +1,142 @@
+            AssertInternable("F#");
+            AssertInternable("VB");
+            AssertInternable("True");
+            AssertInternable("TRUE");
```
Is there a way to ensure they go down the hardcoded path?

ladipro replied:
We could tell by reading the interning stats but I'll first verify that there is value to hardcoding at all. Thanks.

ladipro replied:
I have completely deleted this file, together with the hardcoded string support.

Forgind commented on src/StringTools.UnitTests/OpportunisticIntern_Tests.cs:
```diff
@@ -0,0 +1,142 @@
+
+        private static void AssertInternable(string value)
+        {
+            AssertInternable(value.ToCharArray(), 0, value.ToCharArray().Length);
```
Maybe add a check that the return value equals `value`?

ladipro replied:
Sure, done!

Forgind commented on src/StringTools.UnitTests/OpportunisticIntern_Tests.cs:
```diff
@@ -0,0 +1,142 @@
+        /// </summary>
+        [Fact]
+        public void DoubleDoubleQuotes()
+        {
```
Maybe a test that two strings that were initialized at different points come out to the same reference when interned if they happen to be identical with one (different) string between them that comes out to a different reference?

Like:
```
string foo = "bar";
string quux = "qux";
string baz = "bar";
ReferenceNotEquals(foo, baz); // Have to make them longer for this to be true?
foo = AssertInternable(foo); // Pretending it returns the string
quux = AssertInternable(quux);
baz = AssertInternable(baz);
ReferenceEquals(foo, baz);
ReferenceNotEquals(baz, quux);
```
?

ladipro replied:
`foo` and `baz` in your code snipped are going to be the exact same instance thanks to runtime string interning so the first `ReferenceNotEquals` does not hold, regardless of how long the string is. I believe that a test like you're suggesting - passing two different references to the same contents and verifying that interning them results in the same reference - already exists in `WeakStringCache_Tests`.

Forgind commented on src/StringTools.UnitTests/SpanBasedStringBuilder.Simple_Tests.cs:
```diff
@@ -0,0 +1,142 @@
+using StringToolsNet35::Microsoft.StringTools;
+#endif
+
+namespace Microsoft.StringTools.Tests
```
I don't know if this is possible, but SpanBasedStringBuilder and SpanBasedStringBuilder.Simple are supposed to work the same, right? Is there a way to parameterize a test class and invoke the exact same tests with each?

ladipro replied:
Yes! With just a few more #ifdef's there can be only one SpanBasedStringBuilder_Tests.cs running against both 3.5 and full implementations. I am really glad that you took the time to review this PR, thank you. Fixed.

Forgind commented on src/StringTools.UnitTests/SpanBasedStringBuilder_Tests.cs:
```diff
@@ -0,0 +1,158 @@
+            stringBuilder.Length.ShouldBe(0);
+            stringBuilder.GetEnumerator().MoveNext().ShouldBeFalse();
+        }
+    }
```
It would be nice to have some sort of a performance test. I have a feeling that isn't easily doable here, but it would be nice.

ladipro replied:
I have added a BenchmarkDotNet-based project to the solution.

Forgind commented on src/StringTools.UnitTests/StringTools_Tests.cs:
```diff
@@ -0,0 +1,107 @@
+            InternableString internedString = new InternableString(new string(HardcodedInternedString.ToCharArray()));
+            internedString.ToString().ShouldBeSameAs(HardcodedInternedString);
+
+            InternableString nonInternedString = new InternableString(new string(HardcodedNonInternedString.ToCharArray()));
```
👍 

I really like this test.

Forgind reviewed (approved):
I think this works! Some questions and a couple thoughts on potential points of perf improvement, but this looks good!

ladipro commented on src/Build/Construction/Solution/SolutionProjectGenerator.cs:
```diff
@@ -1097,8 +1099,11 @@ private static string GetConfigurationAndPlatformPropertiesString(ProjectConfigu
-            return directProjectProperties;
+            using SpanBasedStringBuilder directProjectProperties = Strings.GetSpanBasedStringBuilder();
+            directProjectProperties.Append(GetConfigurationAndPlatformPropertiesString(projectConfiguration));
+            directProjectProperties.Append(";");
```
You're right, this is overkill. The advantage of doing it this way is that it's zero-alloc in case the resulting string has already been interned. But it costs significantly more time otherwise and is less readable so I've reverted it, thanks!

ladipro commented on src/Shared/StringToolsHost.cs:
```diff
@@ -0,0 +1,130 @@
+                {
+                    if (candidate[0] == 'C')
+                    {
+                        interned = "C#";
```
String literals are pre-allocated by the runtime and don't cost anything at execution time. Using a literal like this just means that the variable is assigned an object reference - there is no allocation. Strings are even interned by the runtime so if you use the literal `"C#"` in multiple places in the code, they will all refer to the one pre-allocated instance.

Forgind commented on src/Shared/PropertyParser.cs:
```diff
@@ -163,9 +162,22 @@ internal static bool GetTableWithEscaping(TaskLoggingHelper log, string paramete
+                    {
+                        if (needsSemicolon)
+                        {
+                            stringBuilder.Append(";");
```
Looking at your other comments, I'm guessing ";" was pre-allocated (once) by the runtime. Trying to get the hang of that 😄

ladipro replied:
Exactly! Although whether this entire method is now faster than it would be with `String.Join()` (for typical inputs) is an excellent question that deserves careful measurements.

ladipro commented on src/Build/Evaluation/Expander.cs:
```diff
@@ -1318,51 +1334,53 @@ internal static string ConvertToString(object valueToConvert)
                             {
                                 if (builder.Length > 0)
                                 {
-                                    builder.Append(';');
```
In this first iteration I'd like to keep it simple and support only real spans. Making a span out of a character is problematic. `ch.ToString().AsSpan()` allocates so that's a no-go. Maybe the caller could pass the character by-ref so it could be converted to span without allocating, or the `SpanBasedStringBuilder` could keep a list of chars in addition to the list of spans but it feels like it would start defeating the purpose. Good question, though.

ladipro commented on src/Build/Evaluation/Expander.cs:
```diff
@@ -645,68 +644,89 @@ private static string[] ExtractFunctionArguments(IElementLocation elementLocatio
+                    }
+                }
+
+                if (argumentStartIndex.HasValue)
```
The original code was appending to the `argumentBuilder` character by character. With the new data structure this would be inefficient so now it keeps track of the start of the current argument and appends the entire substring when it finds its end. This `if` statement is handling the case where the argument's last character is the last character of the input string so we don't detect the end inside the loop.

ladipro commented on MSBuild.Dev.sln:
```diff
@@ -33,6 +33,12 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Build.Utilities.U
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Xunit.NetCore.Extensions", "src\Xunit.NetCore.Extensions\Xunit.NetCore.Extensions.csproj", "{66C6C1ED-30E7-45DD-9F73-2FBB7889CDBA}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "StringTools", "src\StringTools\StringTools.csproj", "{69FB7694-47E3-45B2-868C-8C3ECB69A947}"
```
Yes to both. VS is handled here and with additional changes in the corresponding insertion. MSBuildLocator will be made aware of the new assembly later.

ladipro commented on src/StringTools/SpanBasedStringBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+            public bool MoveNext()
+            {
+                int newCharIndex = _charIndex + 1;
+                while (_spanIndex < _spans.Count)
```
Not adding empty spans sounds reasonable, done. If spans are guaranteed non-empty, the enumerator can lose one `if` when moving to the next span. I don't think it's worth adding the invariant, especially since maintaining the invariant comes with a cost also. I'll keep as is for now.

ladipro commented on src/StringTools/SpanBasedStringBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+        /// <param name="value">The string to append.</param>
+        public void Append(string value)
+        {
+            AddSpan(value.AsMemory());
```
Done.

ladipro commented on src/StringTools/SpanBasedStringBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+        /// </summary>
+        public void TrimStart()
+        {
+            for (int spanIdx = 0; spanIdx < _spans.Count; spanIdx++)
```
I need the current index to modify the list so rather than a `foreach` + counting the index I believe it's more natural to have a `for` loop and use the same index variable for reading and writing.

ladipro commented on src/StringTools/SpanBasedStringBuilder.cs:
```diff
@@ -0,0 +1,253 @@
+                    _spans[spanIdx] = _spans[spanIdx].Slice(i);
+                    Length -= i;
+                }
+                if (!_spans[spanIdx].IsEmpty)
```
I like the idea. Both `LinkedList` and a circular array look like good options. `LinkedList` can efficiently remove from the middle if we ever support `Remove`. I'd have to keep a pool of nodes, though, to avoid unnecessary allocations. I'll measure and get back to this.

ladipro commented on src/Shared/StringToolsHost.cs:
```diff
@@ -0,0 +1,130 @@
+            }
+            else if (length == 6)
+            {
+                if (TryInternHardcodedString(ref candidate, "''!=''", ref interned) ||
```
Hardcoded strings are not supported anymore, this code is gone.

ladipro commented on src/Shared/StringToolsHost.cs:
```diff
@@ -0,0 +1,130 @@
+        {
+            Debug.Assert(candidate.Length == str.Length);
+
+            if (candidate.StartsWithStringByOrdinalComparison(str))
```
This was supporting strings starting with `"MSBuildProjectInstance"` which were previously hardcoded to **not** be interned. They are unique so if they are really created only once, it is true that interning them is a wasted effort. In this case the effect was negligible and not worth the added complexity, so hardcoded strings, both positive and negative, were removed. If an app/component using `StringTools` doesn't wish to intern certain strings, it should simply not run them through the `Intern` method.

I have converted this method to `Equals`. Thank you!

ladipro commented on src/Shared/StringToolsHost.cs:
```diff
@@ -0,0 +1,130 @@
+                    candidate.StartsWithStringByOrdinalComparison(MSBuildConstants.MSBuildDummyGlobalPropertyHeader))
+            {
+                // don't want to leak unique strings into the cache
+                interned = null;
```
This code was removed.

ladipro commented:
I have finally updated the PR and made it ready for final review. Thank you for bearing with me and for the great feedback so far. Notable changes since the last round:

1. Hardcoded strings are gone. I took measurements without hardcoded strings, with hardcoded strings as they currently exist in the codebase, and with hardcoded strings tailored for a specific scenario (i.e. hardcoding top N strings seen when building a project).
  -- The strings currently in the main branch are not helping. Total time spent interning is actually ~1% worse.
  -- Hardcoding the *right* strings can make things better (by about 1%, interestingly) but the set of *right* strings varies across projects being built. Strings that make one project 1% faster may make little effect on another project or even make it slower. Btw, the most frequently interned strings are not `"true"`, `"false"`, `"=="` but rather things like `"$(MSBuildProjectDirectory)"`, `"@(ReferencePath->'%(ReferenceAssembly)')"`, and `"$([MSBuild]::EnsureTrailingSlash($(MSBuildProjectDirectory)))"`.

   All in all, 1% of string interning cost (not build cost!) is definitely not worth the added complexity and maintenance cost.

2. `SpanBasedStringBuilder` pooling is now done trivially with a thread-static variable rather than with `ObjectPool`. MSBuild does not have code where a given thread would need to hold on to multiple builders at the same time. It's always: 1. Get a builder, 2. Append to it, 3. `ToString()` it. 4. Dispose of it. This makes pooling faster and eliminates the dependency on `Microsoft.Extensions.ObjectPool`.

I've added a simple microbenchmark comparing
- Appending to a regular `StringBuilder` and interning the result of `ToString()`.
- Appending to `SpanBasedStringBuilder` and getting the resulting string with `ToString()` (which has interning built in).
As expected, the results vary wildly depending on how exactly the appending is done - how many strings of what length.

Composing the result out of 1 string with 64 characters is about a wash between the two, both on cache hit (string already interned) and cache miss (string seen for the first time).
In the degenerate case composing the result out of 1 string 1 character long, the regular SB is faster by about 40% on hit and 20% on miss (~200 ns difference in both cases).
On the other end of the spectrum, composing the result out of 16 strings 512 characters each, `SpanBasedStringBuilder` is faster by 15% on hit and by 10% on miss.

In all cases, though, `SpanBasedStringBuilder` is zero-alloc on hit whereas the regular `StringBuilder` allocates the ephemeral string in `ToString`. I'll run a macro-benchmark to quantifying the impact on project evaluation.

@rainersigwald, may I ask you to take a look. Thank you!

danmosemsft commented on src/StringTools/StringTools.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace Microsoft.StringTools
+{
+    public static class Strings
```
Nit re naming, `Strings.Intern(xyz)` might be easily confused (or mistyped) for `String.Intern(xyz)` which is rather different and would cause a leak.

ladipro replied:
Agreed. I have renamed the method to `WeakIntern`.

Forgind commented on src/StringTools/WeakStringCache.Concurrent.cs:
```diff
@@ -0,0 +1,125 @@
+            {
+                // Before we start scavenging set _scavengeThreshold to a high value to effectively lock other threads from
+                // running Scavenge at the same time.
+                if (Interlocked.CompareExchange(ref _scavengeThreshold, int.MaxValue, scavengeThreshold) == scavengeThreshold)
```
I don't think I understand why this works, partially because I would've thought that since the scavengerThreshold is the same as for other threads, if two threads separately decide to run Scavenge, they would both try to set the lock to scavengeThreshold. Am I misunderstanding how Interlocked.CompareExchange works? Or would it be helpful to add a small random number to the scavengeThreshold?

ladipro replied:
If two threads get to this line because the size of the dictionary has exceeded `scavengeThreshold`, they both try to replace `scavengeThreshold` with `int.MaxValue` and the `Interlocked.CompareExchange` call ensures that only one will succeed. That's the one to proceed and run scavenging while the other one will simply skip and not worry about it. When scavenging is in progress, the condition `_stringsByHashCode.Count >= scavengeThreshold` will not hold for anyone and they won't even need the interlocked operation (which is not that cheap either).

It's basically a lock-free implementation of mutual exclusion. We don't want to run scavenging under a lock because we don't want other threads to stop and wait for it to be finished. We just need a lock to "elect" the scavenger thread on contention and this is most efficiently done with this one interlocked operation.

Forgind replied:
I had to read that several times, but I think I get it now. Thanks for the explanation!

ladipro commented:
I have squashed the changes into a reasonable number of commits for easier review.

ladipro commented:
I've benchmarked project evaluation with and without these changes. I have measured evaluation of our own Microsoft.Build.Framework.csproj and of an empty .NET Core console project. Both wall-clock time and memory allocations have improved.

- Memory got better by 0.5% and 2.5%, respectively. These are exact numbers reported by BenchmarkDotNet.
- Time got better by 1% and 6%, respectively. Within error margin so low-confidence numbers, although they correlate with memory nicely.

rainersigwald commented on src/StringTools/WeakStringCache.Concurrent.cs:
```diff
@@ -0,0 +1,124 @@
+                        return result;
+                    }
+
+                    // We have the handle but it's not referencing the right string - create the right string and store it in the handle.
```
what indicates it's not the right string? The hash code matches, so is there an equality check I'm missing?

ladipro replied:
The `GetString` call on line 44 returns non-null only if the stored string matches the incoming `InternableString`.

rainersigwald reviewed (approved):
Many comments, no objections. Looking good!

ladipro commented on src/StringTools/StringTools.csproj:
```diff
@@ -0,0 +1,38 @@
+    <AssemblyName Condition="'$(TargetFramework)' == 'net35'">StringTools.net35</AssemblyName>
+  </PropertyGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' != 'net35'">
```
The previous string interning implementation went to great lengths to also work in 3.5 so I'm afraid that not doing anything at all could lead to impactful regressions. Although I agree that perf regressions may be acceptable, especially when it comes to memory _in one specific process_.

Has the importance of the 3.5 task host gone down compared to, say, five years ago, or is it still about the same?

It's code churn either way - moving the existing 3.5 weak interner into the library or stubbing it out. And since the code already exists and the library may potentially find 3.5 uses outside of MSBuild, I'm leaning towards leaving it in for now.

ladipro commented on src/Package/MSBuild.VSSetup/files.swr:
```diff
@@ -48,6 +48,8 @@ folder InstallDir:\MSBuild\Current\Bin
   file source=$(X86BinPath)System.Runtime.CompilerServices.Unsafe.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Threading.Tasks.Dataflow.dll vs.file.ngenArchitecture=all vs.file.ngenPriority=1
+  file source=$(X86BinPath)StringTools.dll vs.file.ngenArchitecture=all
+  file source=$(TaskHostBinPath)StringTools.net35.dll vs.file.ngenArchitecture=all
```
The .NET 3.5 flavor does not depend on `System.Runtime.CompilerServices.Unsafe`. It's used only in the 4.5/Core implementation.

ladipro commented on src/StringTools/StringTools.csproj:
```diff
@@ -0,0 +1,38 @@
+    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <PreReleaseVersionLabel>preview</PreReleaseVersionLabel>
+
+    <AssemblyName>StringTools</AssemblyName>
```
Renamed, thanks!

ladipro commented on src/StringTools/StringTools.csproj:
```diff
@@ -0,0 +1,38 @@
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+
+    <VersionPrefix>1.0.0</VersionPrefix>
+    <AssemblyVersion>1.0.0.0</AssemblyVersion>
```
Done, thanks!

ladipro commented on src/StringTools.Benchmark/StringTools.Benchmark.csproj:
```diff
@@ -0,0 +1,27 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <Compile Include="..\Shared\FxCopExclusions\Microsoft.Build.Shared.Suppressions.cs">
```
Removed and everything still works, so 👍

ladipro commented on src/StringTools/StringTools.csproj:
```diff
@@ -0,0 +1,38 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Done.

ladipro commented on src/StringTools/InternableString.cs:
```diff
@@ -0,0 +1,331 @@
+        }
+
+        /// <summary>
+        /// The span held by this struct, inline to be able to represent ReadOnlySpan&lt;char&gt;. May be empty.
```
Done.

ladipro commented on src/StringTools/InternableString.cs:
```diff
@@ -0,0 +1,331 @@
+        /// <summary>
+        /// The span held by this struct, inline to be able to represent ReadOnlySpan&lt;char&gt;. May be empty.
+        /// </summary>
+        private ReadOnlySpan<char> _inlineSpan;
```
I thought it wouldn't compile but it does. Fixed, thanks!

ladipro commented on src/StringTools.Benchmark/StringTools.Benchmark.csproj:
```diff
@@ -0,0 +1,27 @@
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Fixed.

ladipro commented on src/StringTools/SpanBasedStringBuilder.Simple.cs:
```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        public SpanBasedStringBuilder(int capacity = 4)
+        {
+            _builder = new StringBuilder(capacity * 128);
```
Done.

ladipro reviewed (commented):
Addressed Rainer's comments and rebased on top of current `master`. If all checks pass, I'll squash the fixup commits into the main work and it will be ready for merge.

ladipro commented:
/azp run

ladipro commented:
Visual Studio insertion with this change will require manual tweaks.
